integration-tests/java/src/integration-test/java/com/stellarstation/api/test/satellite/SatelliteStreamerTest.java

```diff
@@ -55,88 +52,136 @@ void setUp() {
+    class TelemetryAndCommandTestStreamObserver implements StreamObserver<SatelliteStreamResponse> {
+      private final List<Integer> safeModeStates = new ArrayList();
+      private int telemetryReceived;
+      private int totalResponseReceived;
+
+      @Override
+      public void onNext(SatelliteStreamResponse response) {
+        if (response.hasReceiveTelemetryResponse()) {
+          totalResponseReceived++;
+          if (totalResponseReceived > 30) {
+            onCompleted();
+          }
+
+          ByteString data = response.getReceiveTelemetryResponse().getTelemetry().getData();
+          if (data.size() > 2) {
+            // The second last byte of the telemetry indicates the current state of the
+            // fake satellite used in the test. The value is either of 0 or 1.
+            int state = data.byteAt(data.size() - 2);
+            safeModeStates.add(state);
+
+            // Send a command to toggle the state.
+            SendSatelliteCommandsRequest commandsRequest =
+                SendSatelliteCommandsRequest.newBuilder()
+                    .addCommand(ByteString.copyFrom(new byte[] {0x01, 0x01}))
+                    .build();
+            Futures.getUnchecked(requestObserverFuture)
+                .onNext(
+                    SatelliteStreamRequest.newBuilder()
+                        .setSatelliteId(SATELLITE_ID)
+                        .setSendSatelliteCommandsRequest(commandsRequest)
+                        .build());
+
+            telemetryReceived++;
+            if (telemetryReceived == 2) {
+              onCompleted();
+            }
+          }
+        }
+      }
+
+      @Override
+      public void onError(Throwable t) {
+        testCompletionFuture.setException(t);
+      }
+
+      @Override
+      public void onCompleted() {
+        testCompletionFuture.set(true);
+      }
+
+      public List<Integer> getSafeModeStates() {
+        return ImmutableList.copyOf(safeModeStates);
+      }
+    }
+
+    TelemetryAndCommandTestStreamObserver responseObserver =
+        new TelemetryAndCommandTestStreamObserver();
+    StreamObserver<SatelliteStreamRequest> requestObserver = streamer.openStream(responseObserver);
     requestObserverFuture.set(requestObserver);
 
+    requestObserver.onNext(
+        SatelliteStreamRequest.newBuilder().setSatelliteId(SATELLITE_ID).build());
+
+    assertThat(testCompletionFuture.get()).isTrue();
+    assertThat(responseObserver.getSafeModeStates()).containsExactlyInAnyOrder(0, 1);
+  }
+
+  @Test
+  void events() throws Exception {
+    final SettableFuture<Boolean> testCompletionFuture = SettableFuture.create();
+
+    class EventTestStreamObserver implements StreamObserver<SatelliteStreamResponse> {
+      private AntennaState antennaState;
+      private int totalResponseReceived;
+
+      @Override
+      public void onNext(SatelliteStreamResponse response) {
+        totalResponseReceived++;
+        if (totalResponseReceived > 10) {
```

Comment:
If we are only waiting for one thing to be received we could just check the following I think:

if (antennaState != null) {
  return;
}